kani-compiler/src/kani_middle/transform/check_uninit/mod.rs

```diff
@@ -0,0 +1,423 @@
+    ) {
+        let ret_place = Place {
+            local: body.new_local(Ty::bool_ty(), source.span(body.blocks()), Mutability::Not),
+            projection: vec![],
+        };
+        let ptr_operand = operation.mk_operand(body, source);
+        match pointee_info.layout() {
+            PointeeLayout::Sized { layout } => {
+                let is_ptr_initialized_instance = resolve_mem_init_fn(
+                    get_mem_init_fn_def(tcx, "KaniIsPtrInitialized", &mut self.mem_init_fn_cache),
+                    layout.len(),
+                    *pointee_info.ty(),
+                );
+                let layout_operand = mk_layout_operand(body, source, operation.position(), &layout);
+                try_mark_new_bb_as_skipped(&operation, body, skip_first);
+                body.add_call(
+                    &is_ptr_initialized_instance,
+                    source,
+                    operation.position(),
+                    vec![ptr_operand.clone(), layout_operand, operation.expect_count()],
+                    ret_place.clone(),
+                );
+            }
+            PointeeLayout::Slice { element_layout } => {
+                // Since `str`` is a separate type, need to differentiate between [T] and str.
+                let (slicee_ty, diagnostic) = match pointee_info.ty().kind() {
+                    TyKind::RigidTy(RigidTy::Slice(slicee_ty)) => {
+                        (slicee_ty, "KaniIsSlicePtrInitialized")
+                    }
+                    TyKind::RigidTy(RigidTy::Str) => {
+                        (Ty::unsigned_ty(UintTy::U8), "KaniIsStrPtrInitialized")
+                    }
+                    _ => unreachable!(),
+                };
+                let is_ptr_initialized_instance = resolve_mem_init_fn(
+                    get_mem_init_fn_def(tcx, diagnostic, &mut self.mem_init_fn_cache),
+                    element_layout.len(),
+                    slicee_ty,
+                );
+                let layout_operand =
+                    mk_layout_operand(body, source, operation.position(), &element_layout);
+                try_mark_new_bb_as_skipped(&operation, body, skip_first);
+                body.add_call(
+                    &is_ptr_initialized_instance,
+                    source,
+                    operation.position(),
+                    vec![ptr_operand.clone(), layout_operand],
+                    ret_place.clone(),
+                );
+            }
+            PointeeLayout::TraitObject => {
+                try_mark_new_bb_as_skipped(&operation, body, skip_first);
+                let reason = "Kani does not support reasoning about memory initialization of pointers to trait objects.";
+                self.unsupported_check(tcx, body, source, operation.position(), reason);
+                return;
+            }
+        };
+
+        // Make sure all non-padding bytes are initialized.
+        try_mark_new_bb_as_skipped(&operation, body, skip_first);
```

Comment:
nit: A function with `try` in its name means that it is a fallible operation. Something like `collect_skipped` might make more sense.